src/java_bytecode/java_object_factory.cpp

```diff
@@ -28,20 +37,150 @@ Function: gen_nondet_init
 
 \*******************************************************************/
 
-namespace {
-void gen_nondet_init(
+class java_object_factoryt:public messaget
+{
+  code_blockt &init_code;
+  std::set<irep_idt> recursion_set;
+  bool assume_non_null;
+  size_t max_nondet_array_length;
+  symbol_tablet &symbol_table;
+  message_handlert &message_handler;
+  namespacet ns;
+
+public:
+  java_object_factoryt(
+    code_blockt& _init_code,
+    bool _assume_non_null,
+    size_t _max_nondet_array_length,
+    symbol_tablet &_symbol_table,
+    message_handlert &_message_handler):
+    init_code(_init_code),
+    assume_non_null(_assume_non_null),
+    max_nondet_array_length(_max_nondet_array_length),
+    symbol_table(_symbol_table),
+    message_handler(_message_handler),
+    ns(_symbol_table)
+    {}
+
+  exprt allocate_object(
+    const exprt&,
+    const typet&,
+    const source_locationt &,
+    bool create_dynamic_objects);
+
+  void gen_nondet_array_init(const exprt &expr, const source_locationt &);
+
+  void gen_nondet_init(
+    const exprt &expr,
+    bool is_sub,
+    irep_idt class_identifier,
+    const source_locationt &loc,
+    bool skip_classid,
+    bool create_dynamic_objects,
+    bool override=false,
+    const typet &override_type=empty_typet());
+};
+
+
+/*******************************************************************\
+
+Function: gen_nondet_array_init
+
+  Inputs: the target expression, desired object type, source location
+          and Boolean whether to create a dynamic object
+
+ Outputs: the allocated object
+
+ Purpose: Returns false if we can't figure out the size of allocate_type.
+          allocate_type may differ from target_expr, e.g. for target_expr
+          having type int* and allocate_type being an int[10].
+
+\*******************************************************************/
+
+exprt java_object_factoryt::allocate_object(
+  const exprt &target_expr,
+  const typet &allocate_type,
+  const source_locationt &loc,
+  bool create_dynamic_objects)
+{
+  const typet &allocate_type_resolved=ns.follow(allocate_type);
+  const typet &target_type=ns.follow(target_expr.type().subtype());
+  bool cast_needed=allocate_type_resolved!=target_type;
+  if(!create_dynamic_objects)
+  {
+    symbolt &aux_symbol=new_tmp_symbol(symbol_table);
+    aux_symbol.type=allocate_type;
+    aux_symbol.is_static_lifetime=true;
+
+    exprt object=aux_symbol.symbol_expr();
+    exprt aoe=address_of_exprt(object);
+    if(cast_needed)
+      aoe=typecast_exprt(aoe, target_expr.type());
+    code_assignt code(target_expr, aoe);
+    code.add_source_location()=loc;
+    init_code.copy_to_operands(code);
+    return aoe;
+  }
+  else
+  {
+    // build size expression
+    exprt object_size=size_of_expr(allocate_type, namespacet(symbol_table));
+
+    if(allocate_type.id()!=ID_empty && !object_size.is_nil())
```

Comment:
In Java, is it possible that the object size cannot be determined (and size_of_expr would thus return nil)? If that isn't possible, I'd add an assertion instead of the condition.